### PR TITLE
SAK-47247 Send a reminder to students who haven't submitted an assignment

### PR DESCRIPTION
**Pull request by druiz309**
Before, when an instructor creates an assignment and checks the option that allows to send a reminder to students 24 hours before the due date, it creates confusion due to the fact that it also sends a reminder to students who have submitted an assignment previously, So in order to accomplish the specs described on this Jira, I have found where the method that sends the reminder is called and I have filtered members by the ones who haven't submitted the assignment before.

### CODE REVIEW

**Comment by mpellicer** [comment]
I'm confused about this, the method states that sends an email to students who have not submitted the assignment:

https://github.com/sakaiproject/sakai/blob/d5606a5b458d91784e79bf0bfb7390769b899e16/assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java#L85

Is this true or sends the email to everyone in the site?

**Comment by druiz309** [comment]
> 

It seems that sends the email to everyone in the site.

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java [hunk 1]
```diff
@@ -139,10 +141,31 @@ public void execute(String assignmentId) {
...
+                for (int j = 0; j < submitters.size(); j++) {
+                    if(!membersList.get(i).getUserId().equals(submitters.get(j))) {
+                        membersNotSubmitters.add(membersList.get(i));
+                    }
+                }
+            }
+            
             // Do not send reminders if the site is unpublished or softly deleted
             if (site.isPublished() && !site.isSoftlyDeleted()) {
-                for (Member member : site.getMembers()) {
```
Comment (mpellicer) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java:145]: I'm surprised about this, does this mean that the reminder is not group aware??

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java [hunk 2]
```diff
@@ -139,10 +141,31 @@ public void execute(String assignmentId) {
...
+                    .flatMap(s -> s.getSubmitters().stream())
+                    .map(AssignmentSubmissionSubmitter::getSubmitter)
+                    .collect(Collectors.toList());
+            
+            // Get all members in a List collection
+            List<Member> membersList = site.getMembers().stream().collect(Collectors.toList());
+            // List with members filtered (members who are not submitters)
+            List<Member> membersNotSubmitters = new ArrayList<>();
+            
+            for (int i = 0; i < membersList.size(); i++) {
```
Comment (mpellicer) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java:158]: I believe we can rewrite this using streams.

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java [hunk 3]
```diff
@@ -139,10 +141,31 @@ public void execute(String assignmentId) {
...
+                        membersNotSubmitters.add(membersList.get(i));
+                    }
+                }
+            }
+            
             // Do not send reminders if the site is unpublished or softly deleted
             if (site.isPublished() && !site.isSoftlyDeleted()) {
-                for (Member member : site.getMembers()) {
+                for (Member member : membersNotSubmitters) {
                     if (member.isActive() && assignmentService.canSubmit(assignment, member.getUserId()) && !assignmentService.allowAddAssignment(assignment.getContext(), member.getUserId()) && checkEmailPreference(member)) {
```
Comment (mpellicer) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java:169]: canSubmit is just a permission check or checks the submission?

**Review by mpellicer — CHANGES_REQUESTED** [review]
I'm not confident with this approach, maybe this only affects 20.x and it was fixed for future versions of Sakai, could you explore please?

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java (hunk 3 above)
Comment (ottenhoff) [reply assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java:169]: canSubmit is the magic method that should do filtering:

1) does the user have permission
2) does the user have ability to submit a new submission (e.g., has not used all their submissions)
3) if group submission, is a member of the proper group

Please see old JIRAs like:

* SAK-43674 Assignment reminder emails will never send (#8236)
* SAK-42667 Assignments: don't send reminder to instructors (#7486)
* SAK-42380 Assignment canSubmit will properly check permissions (#7401)
* SAK-41865 only send mail to an active member (#6972)
* SAK-40544 add 24 hour reminder email to students for assignments (#5942)

Comment (mpellicer) [reply assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java:169]: Thanks for the detailed info @ottenhoff, you resolved all the investigation I told @druiz309 to do hehe 💯 

A 20.x client complains about this but I believe this problem does not happen in 20.x anymore, @druiz309 is investigating the issue.

Thanks again

**Comment by mpellicer** [comment]
This does not affect 20.x nor newer versions of Sakai, closing it. Thanks for the research.

**Pull request closed without merging**